x/bor/keeper/keeper.go

```
@@ -0,0 +1,353 @@
+func (k *Keeper) AddNewRawSpan(ctx sdk.Context, span hmTypes.Span) error {
+	store := ctx.KVStore(k.storeKey)
+	out, err := k.cdc.MarshalBinaryBare(&span)
+	if err != nil {
+		k.Logger(ctx).Error("Error marshalling span", "error", err)
+		return err
+	}
+	store.Set(GetSpanKey(span.ID), out)
+	return nil
+}
+
+// GetSpan fetches span indexed by id from store
+func (k *Keeper) GetSpan(ctx sdk.Context, id uint64) (*hmTypes.Span, error) {
+	store := ctx.KVStore(k.storeKey)
+	spanKey := GetSpanKey(id)
+
+	// If we are starting from 0 there will be no spanKey present
+	if !store.Has(spanKey) {
+		return nil, errors.New("span not found for id")
+	}
+
+	var span hmTypes.Span
+	if err := k.cdc.UnmarshalBinaryBare(store.Get(spanKey), &span); err != nil {
+		return nil, err
+	}
+
+	return &span, nil
+}
+
+func (k *Keeper) HasSpan(ctx sdk.Context, id uint64) bool {
+	store := ctx.KVStore(k.storeKey)
+	spanKey := GetSpanKey(id)
+	return store.Has(spanKey)
+}
+
+// GetAllSpans fetches all indexed by id from store
+func (k *Keeper) GetAllSpans(ctx sdk.Context) (spans []*hmTypes.Span) {
+	// iterate through spans and create span update array
+	k.IterateSpansAndApplyFn(ctx, func(span hmTypes.Span) error {
+		// append to list of validatorUpdates
+		spans = append(spans, &span)
+		return nil
+	})
+
+	return
+}
+
+// GetSpanList returns all spans with params like page and limit
+func (k *Keeper) GetSpanList(ctx sdk.Context, page uint64, limit uint64) ([]*hmTypes.Span, error) {
+	store := ctx.KVStore(k.storeKey)
+
+	// create spans
+	var spans []*hmTypes.Span
+
+	// have max limit
+	if limit > 20 {
+		limit = 20
+	}
+
+	// get paginated iterator
+	iterator := hmTypes.KVStorePrefixIteratorPaginated(store, SpanPrefixKey, uint(page), uint(limit))
+
+	// loop through validators to get valid validators
+	for ; iterator.Valid(); iterator.Next() {
+		var span hmTypes.Span
+		if err := k.cdc.UnmarshalBinaryBare(iterator.Value(), &span); err == nil {
```

should we return an error if `err != nil`

---

in the else clause, I mean.